### PR TITLE
Initial commit for mixtral on tt-transformer

### PR DESCRIPTION
**Pull request by arekay-tt**
### Ticket
Link to Github Issue

### Problem description
This PR aims to consolidate Mixtral from a separate package to run using the tt-transformer package. This would allow us to refactor the API to allow more models to be executed seamlessly. 

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/tt_transformers/tt/decoder.py [hunk 1]
```diff
@@ -126,22 +161,34 @@ def forward(
...
+        # Check the input sizes here and make sure they are what a MOE expects for Mixtral
+        ff_out = self.feed_forward.forward(ff_in, mode)  # ff_out is replicated
+        ff_out = ff_out[:, :, :, 0:512]
+        # ff_out = ff_out.to_memory_config(memory_config=ttnn.MemoryConfig(memory_config=ttnn.MemoryConfig(memory_layout=ttnn.TensorMemoryLayout.INTERLEAVED,buffer_type=ttnn.BufferType.DRAM)))
+        ff_out = ttnn.to_memory_config(
+            ff_out,
+            memory_config=ttnn.MemoryConfig(
+                memory_layout=ttnn.TensorMemoryLayout.INTERLEAVED, buffer_type=ttnn.BufferType.DRAM
+            ),
+        )
```
Comment (arekay-tt) [inline models/tt_transformers/tt/decoder.py:187]: This is the change to make `ff_out` replicated. Two questions here - is there a better way to do this (convert a replicated to sharded). And secondly, should this be done in the MoE layer or here with a guard via `self.args.is_mixture_of_experts` ?

File: tt_metal/impl/device/device_pool.cpp [hunk 2]
```diff
@@ -263,7 +263,7 @@ void DevicePool::initialize(
 
     _inst->add_devices_to_pool(device_ids);
     _inst->init_firmware_on_active_devices();
-
+    _inst->wait_for_fabric_router_sync();
```
Comment (arekay-tt) [inline tt_metal/impl/device/device_pool.cpp:266]: Not sure if this is the right fix, but was seeing hangs and on debugging i encountered a read waiting inside the next call. My hand-wavy explanation is that the devices have not initialized and hence the next call stalls. With this in place, i have not seen the hang.

Comment (tt-asaigal) [reply tt_metal/impl/device/device_pool.cpp:266]: This should be resolved on main, please rebase with this change removed to double check.

Comment (tt-aho) [reply tt_metal/impl/device/device_pool.cpp:266]: I don't understand this change. What tests were you seeing that were hanging? A test that uses fabric?

Comment (arekay-tt) [reply tt_metal/impl/device/device_pool.cpp:266]: When attempting to run mixtral on T3K with BS>1, the code would hang and attaching/dumping the stack would point to the next line stalling on a read. Specifically `pytest models/tt_transformers/demo/simple_text_demo.py -k "performance and batch-32"` is what is getting stuck.

File: models/tt_transformers/tt/mixtral_mlp.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        super().__init__()
+
+        self.state_dict = state_dict
+        self.mesh_device = mesh_device
+        self.dtypes = dtypes
+        self.model_args = args
+        self.model_config = args.get_model_config()
+
+        # Porting mixtral to llama
+        self.model_config["FF3_OUTPUT_PROGCFG"] = ttnn.MatmulMultiCoreReuseMultiCast1DProgramConfig(
```
Comment (yieldthought) [inline models/tt_transformers/tt/mixtral_mlp.py:22]: Much as I dislike model_config, I think it's very unexpected to assign to its variables here. These program configs should not be hard-coded, they should be generated using the model_config helper function matmul_1d_config_from_tensor_shapes. If the issue is that the Mixtral MLP is not tensor-parallel across devices then that should be handled with an `if is_moe` or whatever in the place these variables are normally assigned in the ModelArgs constructor.

File: models/tt_transformers/tt/mixtral_mlp.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            )
+            w2_in = ttnn.mul(w1_out, w3_out)
+            w1_out.deallocate(True)
+            w3_out.deallocate(True)
+            w2_out = ttnn.matmul(
+                w2_in,
+                self.w2,
+                program_config=self.model_config["FF2_OUTPUT_PROGCFG"],
+                memory_config=self.model_config["FF2_OUTPUT_MEMCFG"],
+                compute_kernel_config=ttnn.WormholeComputeKernelConfig(
```
Comment (yieldthought) [inline models/tt_transformers/tt/mixtral_mlp.py:219]: Do not create compute kernel configs in every forward pass, use a predefined one from ModelArgs e.g. this code from mlp.py:
```
                self.args.compute_kernel_config_lofi
                if self.four_bit_mlp
                else self.args.compute_kernel_config_hifi2_fp16
```

File: models/tt_transformers/tt/mixtral_moe.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        self.dtype = dtype
+        self.model_config = args.get_model_config()
+
+        gate_name = f"layers.{layer_num}.feed_forward.gate.weight"
+        if args.dummy_weights:
+            cache_name = None
+        else:
+            cache_name = args.weight_cache_path(dtype) / (gate_name + "_multidevice_repadded")
+
+        self.num_devices = 8
```
Comment (yieldthought) [inline models/tt_transformers/tt/mixtral_moe.py:26]: What if we do not have 8 devices? model_config knows how many devices there are.

File: models/tt_transformers/tt/mixtral_moe.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            torch.cat(gates_tensor_list, dim=1),
+            dtype=ttnn.bfloat16,
+            layout=self.model_config["GATE_W_LAYOUT_TILE"],
+            memory_config=self.model_config["GATE_WEIGHTS_MEMCFG"],
+            cache_file_name=cache_name,
+            device=self.mesh_device,
+            mesh_mapper=ShardTensorToMesh(mesh_device, dim=1),
+        )
+
+        self.tile_size = 32
```
Comment (yieldthought) [inline models/tt_transformers/tt/mixtral_moe.py:50]: Already defined in `args.tile_size`

File: models/tt_transformers/tt/mixtral_moe.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        # self.compute_kernel = args.get_compute_kernel_attn_config()
```
Comment (yieldthought) [inline models/tt_transformers/tt/mixtral_moe.py:51]: Do not merge commented out code

File: models/tt_transformers/tt/mixtral_moe.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        # self.compute_kernel_reduce = args.get_compute_kernel_config_reduce()
+        # self.compute_kernel = args.get_compute_kernel_attn_config()
+        self.compute_kernel = ttnn.WormholeComputeKernelConfig(
```
Comment (yieldthought) [inline models/tt_transformers/tt/mixtral_moe.py:54]: compute kernel configs are predefined in ModelArgs ctor, use those and add one if you have to.

File: models/tt_transformers/tt/mixtral_moe.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        )
+        # self.compute_kernel_reduce = args.get_compute_kernel_config_reduce()
+        self.compute_kernel_reduce = ttnn.WormholeComputeKernelConfig(
+            math_fidelity=ttnn.MathFidelity.HiFi2,
+            fp32_dest_acc_en=False,
+            packer_l1_acc=False,
+        )
+
+        top8_mask = torch.full((1, 1, 1, 64), fill_value=torch.finfo(torch.float).min)
+        top8_mask[:, :, :, :8] = 0.0
```
Comment (yieldthought) [inline models/tt_transformers/tt/mixtral_moe.py:67]: If you don't want to generalize to situations other than num_devices = num_experts = 8 then add an assert for this.

File: models/tt_transformers/tt/mixtral_moe.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+            output_11BH_gathered.deallocate(True)
+        else:  # Decode mode
+            output_11BH_gathered = ttnn.all_gather(results_11BH, dim=2, num_links=1)
+            results_11BH.deallocate(True)
+            # Reduction
+            output_11BH_reduced = ttnn.matmul(
+                self.reduce_mask, output_11BH_gathered, compute_kernel_config=self.compute_kernel_reduce
+            )
+
+        def replicate_to_shard(tensor):
```
Comment (yieldthought) [inline models/tt_transformers/tt/mixtral_moe.py:160]: Defining this function inline and then calling it exactly once doesn't do anything readability and conciseness

File: models/tt_transformers/tt/model_config.py [hunk 11]
```diff
@@ -558,6 +562,23 @@ def __init__(
             )
 
             # Configure data precision and math fidelity for tensors and kernels
+
+            self.model_config["COMPUTE_KERNEL_CONFIG_HIFI2"] = self.compute_kernel_config_hifi2
+
+            # Mixtral
+            self.model_config["PREFILL_MLP_COMPUTE_CONFIG"] = ttnn.WormholeComputeKernelConfig(
```
Comment (yieldthought) [inline models/tt_transformers/tt/model_config.py:569]: Why not `self.compute_kernel_config_lofi`?

File: models/tt_transformers/tt/model_config.py [hunk 12]
```diff
@@ -558,6 +562,23 @@ def __init__(
...
+                math_fidelity=ttnn.MathFidelity.LoFi,
+                math_approx_mode=True,
+                fp32_dest_acc_en=False,
+                packer_l1_acc=True,
+            )
+            self.model_config["GATE_MM_OUTPUT_KERNEL_CONFIG"] = ttnn.WormholeComputeKernelConfig(
```
Comment (yieldthought) [inline models/tt_transformers/tt/model_config.py:575]: Why not `self.compute_kernel_config_hifi4`?

**Review by yieldthought — CHANGES_REQUESTED** [review]

File: models/tt_transformers/tt/common.py [hunk 13]
```diff
@@ -118,11 +118,11 @@ def preprocess_inputs_prefill(
     decoding_pos = []
     prefill_lens = []
 
-    # Always prefill the nearest power of 2 for each user. This means that the majority of cases we will prefill more tokens than needed.
+    prefill_seq_len = max_prompt_len
+    # Always prefill to max prompt len so we can stack them together
     # To avoid issues, we keep track of the decoding position to decode correctly the user's prompt
     for i, encoded in enumerate(encoded_prompts):
         # Prefill size is nearest power of 2
-        prefill_seq_len = max(2 ** math.ceil(math.log(len(encoded), 2)), 128)
```
[the remark this answers is not in the document]
Comment (arekay-tt) [reply models/tt_transformers/tt/common.py:125]: The tensors are stacked together and need to be the same size. While we prefill one request at a time, we do pass them in as a single tensor.

File: models/tt_transformers/tt/decoder.py [hunk 14]
```diff
@@ -127,15 +153,20 @@ def forward(
...
             chunk_start_idx=chunk_start_idx,
             kv_cache=kv_cache,
-        )
+        )  # attn_out is sharded across devices,
+
+        # TODO - save a copy in torch and then perform the computation on the side
+        # to check if this is accurate to what we are expecting.
         # Here x and attn_out are both fractured across devices
-        h = ttnn.add(x, attn_out, memory_config=skip_mem_cfg, dtype=ttnn.bfloat16 if TG else None)
+        h_val = ttnn.add(
```
[the remark this answers is not in the document]
Comment (arekay-tt) [reply models/tt_transformers/tt/decoder.py:159]: `pdb` has `h` as the help command so it (or maybe just me) gets confused.

File: models/tt_transformers/tt/mixtral_moe.py (hunk 10 above)
Comment (arekay-tt) [reply models/tt_transformers/tt/mixtral_moe.py:151]: The primary intent here is to not expose this outside since i am not sure if the implementation covers all corner cases and is intended for the specific use given the context. I do think that having a name for the block helps readability but open to suggestions.

**Comment by uaydonat** [comment]
There is urgency to get this model in. @yieldthought  If the comments are about code cleanliness and there are no fundamental issues with the PR, we should approve it.

File: models/tt_transformers/demo/simple_text_demo.py [hunk 15]
```diff
@@ -586,7 +586,7 @@ def test_demo_text(
         logger.info("Starting prefill warmup...")
         profiler.start(f"compile_prefill", iteration=batch_idx)
         logits = generator.prefill_forward_text(
-            input_tokens_prefill_pt[::batch_size, :],  # Warmup prefill for each device
+            input_tokens_prefill_pt[:num_devices, :],  # Warmup prefill for each device
```
Comment (yieldthought) [inline models/tt_transformers/demo/simple_text_demo.py:589]: I'm very confused by this, ::batch_size is taking every token with BS=1 or every 2nd for BS=2 etc. :num_devices takes the first N. Was the original just completely wrong?

Comment (yieldthought) [reply models/tt_transformers/demo/simple_text_demo.py:589]: I think this change breaks data-parallel support.

Comment (arekay-tt) [reply models/tt_transformers/demo/simple_text_demo.py:589]: If i understand correctly, the original version is trying to sample the inputs to get warmup for each batch (first request in each batch). The mismatch is that the second request in the prefill takes significantly longer than others unless we run prefill on it as well. 

It shouldn't break things as it is an optimization - it just warms up the prefill path and improves TTFT unless there are some side-effects.

**Review by yieldthought — APPROVED** [review]
Pre-approved after discussion of changes to come

**Comment by BossChaos** [comment]
## Code Review

**PR:** [Initial commit for mixtral on tt-transformer](https://github.com/tenstorrent/tt-metal/pull/19398) by @arekay-tt

- ✅ Code review

**Wallet:** `0xdaE5d307339074A24F579dB48e7c639359D94904`

*Code review for Tenstorrent tt-metal — bounty #8621*
